apps/dashr-pk-calc/app.R

```diff
@@ -0,0 +1,337 @@
+               input(id = "subjects-input", property = "value"),
+               state(id = "data-table", property = "data")
+             ),
+             
+  function(rows, subjects, records){
+   
+   changeRow <- rows - length(records)
+
+   if (changeRow > 0) {
+     for (i in (length(records) +1):(length(records) + changeRow)){
+     #iterate through new rows
+       
+       records[[i]] <- records[[i-1]]
+       # Copy previous row content to new
+       records[[i]][1:length(records[[i]])] <- ""
+       # Replace with empty string
+     }
+     
+   }else if (changeRow < 0){
+     records <- records[1:rows]
+   }
+   
+   recordDf <- rbindlist(records) %>% select(time, everything())
+   # Create df from records time first place
+
+   changeCol <- subjects - (ncol(recordDf) -1)
+   
+   if (changeCol > 0) {
+     
+     numLengths <- as.numeric(names(recordDf)[2:length(recordDf)])
+     nameAppend <- numLengths[length(numLengths)] +1
+     # Get name for new column
+     recordDf <- cbind(recordDf, "")
+     # Create new column
+     names(recordDf)[ncol(recordDf)] <- nameAppend
+     # Set the name
+   }else if (changeCol < 0){
+     recordDf <- select(recordDf, 1:(subjects+1))
+     # Remove columns
+   }
+   
+   records <- df_to_list(recordDf)
+   # convert df back to list
+   return(records)
+  }
+)
+
+app$callback(output = list(id = "results-table", property = "columns"),
+             params = list(
+               input(id = "data-table", property = "data")
+             ),
+             
+  function(results){
+    return(GenerateResultColumn(results))
+  }
+)
+
+app$callback(output = list(id = "results-table", property = "data"),
+             params = list(
+               input(id = "data-table", property = "data")
+             ),
+
+  function(records){
+    results <- GenerateResultData(records)
+  }
+)
+
+app$callback(output = list(id = "results-graph", property = "figure"),
+             params = list(
+               input(id = "data-table", property = "data")
+             ),
+  function(records){
+    
+    df <- rbindlist(records)
+    # Convert records to df
+    
+    df <- as.data.frame(sapply(df, function(x) as.numeric(as.character(x))))
+    # Convert anything non-numeric to NA
+    
+    df <- df[,colSums(is.na(df)) < nrow(df)]
+    # Remove column if all NA
+    
+    dfMelt <- melt(df, id.vars = "time")
+    names(dfMelt) <- c("x", "name", "y")
+    dfMelt <- dfMelt[,c("x", "y", "name")]
+    dfMelt <- mutate(dfMelt, mode = "lines+markers")
+    # Reformat df to easily feed into list
+
+    figData <- list(list())
+    # Initiate figData list
+    for (i in 1:length(unique(dfMelt$name))){
+      dfSub <- dfMelt[dfMelt$name == as.character(i-1),]
+      figData[[i]] <- list(
+        "x" = as.numeric(dfSub$x),
+        "y" = as.numeric(dfSub$y),
+        "name" = paste("Subj", i, sep = ""),
+        "mode" = "lines+markers"
+      )
+    }
+    # Filled the figData list
```

Comment:
Probably just a personal preference to avoid explicit `for` loops whenever I can, but I would rewrite this chunk as something like:
```
    figData <- lapply(
      1:length(unique(dfMelt$name)),
      function(i){
        dfSub <- dfMelt[dfMelt$name == as.character(i-1),]
        list(
          x = as.numeric(dfSub$x),
          y = as.numeric(dfSub$y),
          name = paste("Subj", i, sep = ""),
          mode = "lines+markers"
        )
      }
    )
```
But it probably doesn't make much of a difference.

Reply:
I agree, I think it makes for cleaner code. This one is close enough for now, I think we can leave as-is and revisit.